### DISCUSSION TITLE
micropy/SX126X (LoRa) memory allocation error on PyBoard but not Pico

### DISCUSSION THREAD

**Discussion by MarkMakies** [micropython/micropython#15180] [STM32 / Pyboard] [answered]
Whilst trying to import SX126X LoRa module drivers [https://github.com/ehong-tl/micropySX126X] into an original PyBoard v1.1 I get a memory allocation error.  Previously I had no problems with a RP2040 based Pico with this same LoRa module.  I'm guessing it is because Pico has 264k whilst the PyB has 192k.

It doesn't sound quite right to me that the reference board doesn't have enough RAM to compile this module.  Could this be the case?  There is no other code in main.py and the boards were reset prior to test.

**MicroPython v1.22.2 (double precision version) on 2024-02-22; PYBv1.1 with STM32F405RG**

```
>>> import micropython
>>> import gc
>>> 
>>> gc.collect()
>>> print(micropython.mem_info())
stack: 500 out of 15360
GC: total: 103168, used: 1472, free: 101696
 No. of 1-blocks: 17, 2-blocks: 5, max blk sz: 40, max free sz: 6342
None
>>> print(gc.mem_free())
101504

>>> from sx1262 import SX1262
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "sx1262.py", line 2, in <module>
MemoryError: memory allocation failed, allocating 640 bytes
```

**MicroPython v1.20.0 on 2023-04-26; Raspberry Pi Pico with RP2040**
```
>>> import micropython
>>> import gc
>>> 
>>> gc.collect()
>>> print(micropython.mem_info())
stack: 508 out of 7936
GC: total: 226560, used: 4784, free: 221776
 No. of 1-blocks: 40, 2-blocks: 16, max blk sz: 64, max free sz: 13803
None
>>> print(gc.mem_free())
221584

>>> from sx1262 import SX1262
>>> 


```

**Comment by robert-hh** [comment, marked as answer]
You could try to pre-compile the module using mpy-cross. That save a little bit of RAM during import. The other option is to embed the LoRa module as frozen bytecode, requiring to build you own version of the firmware. Then the driver code remains in flash and does not use RAM.

Note: PyBoard was the first device used for implementing MicroPython more then 10 years ago. If the RP29040 would have existed at that time, it might have been chosen instead as a platform. Today, PybBoard D devices are the better reference in the PyBoard range of boards.

**Reply by MarkMakies** [reply]
Thanks Robert.  The pre-compile worked perfectly.  I haven't tried building my own firmware yet, but I'll give it a go.

I recently got a few PyBorads, thinking that they would be a good reference, after working mainly with ESP32 and RP2024.  I also needed a breakout board that could concurrently supports 4x UARTS, 1x SPI and 1x I2C. Looks like I've chosen the wrong board as the D can do all of this.  Thanks for the advice.
